ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,190 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        FullInfo,
+                        Size,
+                        &Size);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Enumerate subkeys (i.e. the different device objects) */
+    for (i = 0; i < FullInfo->SubKeys; i++)
+    {
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                NULL,
+                                0,
+                                &Size);
+        if (Status == STATUS_NO_MORE_ENTRIES)
+        {
+            break;
+        }
+        else if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        BasicInfo = ExAllocatePoolWithTag(PagedPool,
+                                          Size,
+                                          TAG_IO);
+        if (!BasicInfo)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto cleanup;
+        }
+
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                BasicInfo,
+                                Size,
+                                &Size);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        /* Check BasicInfo->Name */
+        RtlInitUnicodeString(&BasicInfoName, NULL);
+        RtlInitUnicodeString(&InLinkName, NULL);
+
+        BasicInfoName.Length = wcslen(BasicInfo->Name);
+        BasicInfoName.MaximumLength = wcslen(BasicInfo->Name);
+        BasicInfoName.Buffer = BasicInfo->Name;
+
+        AliasKeyName.Length = 0;
+        AliasKeyName.MaximumLength += BasicInfoName.Length;
+        AliasKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
+                                                    AliasKeyName.MaximumLength,
+                                                    TAG_IO);
+        if (!AliasKeyName.Buffer)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto cleanup;
+        }
+
+        Status = RtlAppendUnicodeStringToString(&AliasKeyName, &BasicInfoName);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        pdest = wcsstr(SymbolicLinkName->Buffer, L"Services\\");
```

Comment:
`UNICODE_STRING` buffers are not guaranteed to be null-terminated, so you can't use `wcsstr` on them.
If you know that this string is always supposed to start with a certain prefix, use `RtlPrefixUnicodeString` to verify, then use a fixed offset into the string.
If not, you may have to make a copy of the string and ensure that that's null-terminated. Or, since I think this function is supposed to receive a string that another Io function returned, perhaps we can just fail in that case?

Reply:
> use a fixed offset into the string.

Sorry, but how I can use it? Which function or macro should I call for it?

> we can just fail in that case?

Not necessarily. I think it may fail when the prefix is not found, but it supposed to be found in `SymbolicLinkName`, if the appropriate parameter of the call is specified correctly.